### PR TITLE
improve: added pagination model and selector for future use

### PR DESCRIPTION
**Pull request by j-sneh**
## Summary

Created the PaginationResponse, PaginationRequest, and internal Pagination model as an incremental change toward fully-integrated pagination.

## Checklist

~~- [ ] Wrote appropriate unit tests~~
- [x] Considered security implications of the change
~~- [ ] Updated associated docs where necessary~~
~~- [ ] Updated associated configuration where necessary~~
- [x] Change is backwards compatible if it needs to be (user can upgrade without manual steps?)
- [x] Nothing sensitive logged
- [x] Considered data migrations for smooth upgrades

### CODE REVIEW

File: internal/server/models/pagination.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"strings"
+
+	"github.com/infrahq/infra/api"
+)
+
+// Internal Pagination Data
+type Pagination struct {
+	Page    int
+	Limit   int
+	Sort    string
```
Comment (jmorganca) [inline internal/server/models/pagination.go:14]: Should this change include `Sort` ?

File: internal/server/models/pagination.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+func (p *Pagination) SetCount(count int64) {
+	p.MaxPage = int(math.Ceil(float64(count) / float64(p.Limit)))
+	p.Total = int(count)
+
+	if p.Next > p.MaxPage {
+		p.Next = 0
+	}
+}
+
+func (p *Pagination) SetDefaultSort(sort string) {
```
Comment (jmorganca) [inline internal/server/models/pagination.go:79]: Similar to above: do we want to include for this change, or in a future change?

File: internal/server/models/pagination.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/infrahq/infra/api"
+)
+
+// Internal Pagination Data
+type Pagination struct {
+	Page    int
+	Limit   int
+	Sort    string
+	Total   int
+	MaxPage int
```
Comment (jmorganca) [inline internal/server/models/pagination.go:16]: Is this the total # of pages? We could use `Pages` instead

File: internal/server/models/pagination.go (hunk 2 above)
Comment (j-sneh) [reply internal/server/models/pagination.go:79]: Why push off sorting to a future change?

**Review by BruceMacD — COMMENTED** [review]
Good step 👍

File: internal/server/data/selectors.go [hunk 4]
```diff
@@ -118,6 +119,18 @@ func ByUserID(userID uid.ID) SelectorFunc {
 	}
 }
 
+func ByPagination(pg models.Pagination) SelectorFunc {
```
Comment (jmorganca) [inline internal/server/data/selectors.go:122]: Would it be possible to add a few tests for this selector?

File: internal/server/models/pagination.go [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+		Page:  page,
+		Limit: limit,
+		Sort:  sort,
+		Prev:  prevPage,
+		Next:  page + 1,
+	}
+}
+
+// PaginationToResponse converts the internal Pagination model to a response sent to the user
+func (p *Pagination) PaginationToResponse() api.PaginationResponse {
```
Comment (jmorganca) [inline internal/server/models/pagination.go:55]: Similar to above comment: consider this in a follow up PR

File: internal/server/models/pagination.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+// Internal Pagination Data
+type Pagination struct {
+	Page  int
+	Limit int
+	Sort  string
+	Total int
+	Pages int
+	Next  int
+	Prev  int
```
Comment (mxyng) [inline internal/server/models/pagination.go:18]: What are the use cases for these fields?

Comment (j-sneh) [reply internal/server/models/pagination.go:18]: If there is no next or previous page then the field would be empty, other than that it's just a ±1. I guess they're not very necessary

File: internal/server/models/pagination.go [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+	"strings"
+
+	"github.com/infrahq/infra/api"
+)
+
+// Internal Pagination Data
+type Pagination struct {
+	Page  int
+	Limit int
+	Sort  string
```
[the remark this answers is not in the document]
Comment (j-sneh) [reply internal/server/models/pagination.go:11]: I think we should stick with the sorting. It's already working in the data layer, it's relevant to pagination and quite simple (one call to db.Order and some string validation).

**Review by BruceMacD — APPROVED** [review]
This looks good to me once other people follow up on their feedback. Great work.

File: internal/server/models/pagination.go [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+	"math"
+)
+
+// Internal Pagination Data
+type Pagination struct {
+	Page  int
+	Limit int
+	Sort  string
+
+	TotalCount int
```
Comment (jmorganca) [inline internal/server/models/pagination.go:13]: Do we need these fields yet (or the functions below)? It doesn't seem they are called + tested anywhere, and I'd be weary of needing to have "calculated" state. We already have `Count` in our `ListResponse` type.

File: internal/server/models/pagination.go [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+	TotalPages int
+}
+
+func (p *Pagination) SetCount(count int64) {
```
Comment (jmorganca) [inline internal/server/models/pagination.go:17]: Similar to above comment – having to "calculate" state can be a tough developer experience using this struct

File: internal/server/models/pagination.go (hunk 8 above)
Comment (j-sneh) [reply internal/server/models/pagination.go:13]: This is a good point and calculating the total count/pages is less than ideal (querying a whole table twice) and can contribute to some wonky design. We do need some sort of method of communicating to the user how many requests are necessary to get the full list

**Review by jmorganca — APPROVED** [review]

**Pull request merged**